File: java/week_1/BinarySearch.java
import java.util.InputMismatchException;
import java.util.Scanner;

public class BinarySearch {
	static SearchResult binary_search(int[] haystack, int needle) {
		int top = haystack.length - 1;
		int bottom = 0;
		int steps = 0;
		while (bottom <= top) {
			steps++;
			int middle = (top + bottom) / 2;
			int guess = haystack[middle];
			if (guess == needle) return new SearchResult(middle, steps); 
			if (guess > needle) top = middle - 1;
			if (guess < needle) bottom = middle + 1;
		}
		return null;
	}

	static void fill_array(int[] array) {
		for (int i = 0; i < array.length; i++) {
			array[i] = (1+i);
		}
	}

	static int validate_integer_input(int var, Scanner sc) throws IllegalArgumentException {
		try {
			var = sc.nextInt();
			return var;
		} catch (InputMismatchException e) {
			System.out.println("\nInvalid character entered! Please only input integer numbers.");
			throw e;
		}
	}
	
	public static void main(String[] args) {
		Scanner sc = new Scanner(System.in);
		try {
			System.out.print("Welcome! First, please enter the desired var for the array: ");
			int size = 0; size = validate_integer_input(size, sc);
		
			int[] haystack = new int[size];
			fill_array(haystack);
		
			System.out.print("\nOK! Now, enter the element to be looked for: ");
			int needle = 0; needle = validate_integer_input(needle, sc);
			
			SearchResult response = binary_search(haystack, needle);
			if (response == null) {
				System.out.println("The element could not be found! Try again :/");
				return;
			}	
			System.out.printf("\nThe element position is %d, and it was found in %d steps! xD", response.index, response.steps);
		} catch (InputMismatchException e) {
			return;
		} catch (Exception e) {
			System.out.println("\nAn unexpected error happened! xO");
			return;
		} finally {
			sc.close();
		}
	}
}

class SearchResult {
	int index;
	int steps;

	public SearchResult(int index, int steps) {
		this.index = index;
		this.steps = steps;
	}
}

File: algorithms/c/week_1/binary_search.c
#include <stdio.h>
#include <stdlib.h>

struct SearchResult {
	int index;
	int steps;
};

struct SearchResult binary_search(int *haystack, int needle, int size);
void fill_array(int *arr, int arr_size);

int main() {
	printf("Welcome! First, enter the desired size for the array: ");
	int size;
	scanf("%d", &size);

	int *haystack = malloc(size * sizeof(int));

	if (haystack == NULL) {
		printf("Memory allocation failed! Xd\n");
		return 1;
	}

	int needle;

	fill_array(haystack, size);
	printf("Enter a number between 1 and %d: ", size);
	scanf("%d", &needle);
	struct SearchResult response = binary_search(haystack, needle, size);
	if (response.index == -1) { 
		printf("NOT FOUND! The element couldn't be find :/\n Try again later, bye!"); 
		return 0; 
	}
	printf("\nFOUND! The sought element is at position %d, and was found in %d steps! xD", response.index, response.steps);
}

struct SearchResult binary_search(int *haystack, int needle, int size) {
	int bottom = 0;
	int top = size - 1;
	int steps = 0;
	struct SearchResult response;
	while (bottom <= top) {	
		steps++;
		int middle = (top + bottom) / 2;
		int guess = haystack[middle];
		if (needle == guess) {
			response.index = middle;
			response.steps = steps;
			return response;
		};
		if (needle < guess) top = middle - 1;
		if (needle > guess) bottom = middle + 1;
	}
	response.index = -1;
	return response;
}

void fill_array(int *arr, int arr_size) {
	for (int i = 0; i < arr_size; i++) {
		*(arr + i) = (1+i);
	}
}

